# Analyzing 2

In this jupyter notebook the dataframe 'mosquito1_meteo.csv' will be analized. The objectives are:

- Mean meteo data per country obtention.

- Mosquito per day per country obtention.

- Get the dataframe ready for machine learning.

In [92]:
# Data treatment
# ------------------------------------------------------------------------------
import numpy as np
import pandas as pd
#from datetime import date, datetime
#import holidays
#import requests #To check if an API link works
#from IPython.core.interactiveshell import InteractiveShell #Show more than one output per cell
#InteractiveShell.ast_node_interactivity = "all"
#import ast

# API for accessing open weather and climate data
# ------------------------------------------------------------------------------
#from meteostat import Point, Daily

# Graphs
# ------------------------------------------------------------------------------
import matplotlib.pyplot as plt
import seaborn as sns

# Warnings configuration
# ------------------------------------------------------------------------------
#import warnings
#warnings.filterwarnings('once')

# Library to create pickle files.
# ------------------------------------------------------------------------------
#import pickle
#import os

# Progress bar
# ------------------------------------------------------------------------------
#from tqdm import tqdm

# Put a pause between API calls
# ------------------------------------------------------------------------------
#import time

In [93]:
df_1 = pd.read_csv("../data/mosquito1_meteo.csv",index_col=0)
df_1.head(3)

index  event_date  year  month  day country_code  latitude  longitude  \
0      0  2022-11-04  2022     11    4           ES  41.51019    2.24589   
1      1  2021-08-27  2021      8   27           IT  44.40289    8.98775   
2      3  2022-08-11  2022      8   11           IT  41.70922   12.78512   

                                             witness  \
0                                       Roger Eritja   
1  Karin Bakran-Lebl;Ana Klobucar;UNIROMA1;Roger ...   
2  UNIROMA1;Eleonora Longo;Francesco Severini;Rog...   

                                issue  ...  tavg  tmin  tmax  prcp snow  \
0  CONTINENT_DERIVED_FROM_COORDINATES  ...  15.4  11.0  20.0   0.0  NaN   
1  CONTINENT_DERIVED_FROM_COORDINATES  ...  25.0  23.0  28.1   0.3  NaN   
2  CONTINENT_DERIVED_FROM_COORDINATES  ...  27.7  22.0  32.0   1.1  NaN   

    wdir  wspd  wpgt    pres  tsun  
0  271.0  19.5  42.6  1013.8   NaN  
1   68.0  13.8  27.8  1007.5   NaN  
2  173.0   9.0   NaN  1013.9   NaN  

[3 rows x 25 columns]

In [94]:
print(f'This DataFrame has {df_1.shape[0]} rows and {df_1.shape[1]} columns.')

This DataFrame has 14512 rows and 25 columns.


## 1. Clean dataframe

### 1.1. Check concat

In [95]:
# Change the column 'event_date' to datetime format.
df_1['new_event_date'] = pd.to_datetime(df_1['event_date'], format='%Y-%m-%d')
df_1.head(3)

index  event_date  year  month  day country_code  latitude  longitude  \
0      0  2022-11-04  2022     11    4           ES  41.51019    2.24589   
1      1  2021-08-27  2021      8   27           IT  44.40289    8.98775   
2      3  2022-08-11  2022      8   11           IT  41.70922   12.78512   

                                             witness  \
0                                       Roger Eritja   
1  Karin Bakran-Lebl;Ana Klobucar;UNIROMA1;Roger ...   
2  UNIROMA1;Eleonora Longo;Francesco Severini;Rog...   

                                issue  ...  tmin  tmax  prcp  snow   wdir  \
0  CONTINENT_DERIVED_FROM_COORDINATES  ...  11.0  20.0   0.0   NaN  271.0   
1  CONTINENT_DERIVED_FROM_COORDINATES  ...  23.0  28.1   0.3   NaN   68.0   
2  CONTINENT_DERIVED_FROM_COORDINATES  ...  22.0  32.0   1.1   NaN  173.0   

   wspd  wpgt    pres  tsun  new_event_date  
0  19.5  42.6  1013.8   NaN      2022-11-04  
1  13.8  27.8  1007.5   NaN      2021-08-27  
2   9.0   NaN  1013.9   NaN      2022-08-11  

[3 rows x 26 columns]

In [96]:
# Change the column 'event_date' to datetime format.
df_1['new_event_date2'] = pd.to_datetime(df_1['event_date2'], format='%Y-%m-%d')
df_1.head(3)

index  event_date  year  month  day country_code  latitude  longitude  \
0      0  2022-11-04  2022     11    4           ES  41.51019    2.24589   
1      1  2021-08-27  2021      8   27           IT  44.40289    8.98775   
2      3  2022-08-11  2022      8   11           IT  41.70922   12.78512   

                                             witness  \
0                                       Roger Eritja   
1  Karin Bakran-Lebl;Ana Klobucar;UNIROMA1;Roger ...   
2  UNIROMA1;Eleonora Longo;Francesco Severini;Rog...   

                                issue  ...  tmax prcp  snow   wdir  wspd  \
0  CONTINENT_DERIVED_FROM_COORDINATES  ...  20.0  0.0   NaN  271.0  19.5   
1  CONTINENT_DERIVED_FROM_COORDINATES  ...  28.1  0.3   NaN   68.0  13.8   
2  CONTINENT_DERIVED_FROM_COORDINATES  ...  32.0  1.1   NaN  173.0   9.0   

   wpgt    pres  tsun  new_event_date  new_event_date2  
0  42.6  1013.8   NaN      2022-11-04       2022-11-04  
1  27.8  1007.5   NaN      2021-08-27       2021-08-27  
2   NaN  1013.9   NaN      2022-08-11       2022-08-11  

[3 rows x 27 columns]

In [97]:
# Check the type of information that each column has.
df_1.dtypes

index                       int64
event_date                 object
year                        int64
month                       int64
day                         int64
country_code               object
latitude                  float64
longitude                 float64
witness                    object
issue                      object
index.1                     int64
event_date2                object
latitude2                 float64
longitude2                float64
time                       object
tavg                      float64
tmin                      float64
tmax                      float64
prcp                      float64
snow                      float64
wdir                      float64
wspd                      float64
wpgt                      float64
pres                      float64
tsun                      float64
new_event_date     datetime64[ns]
new_event_date2    datetime64[ns]
dtype: object

In [98]:
# Check if column 'new_check_date' has the same information as 'new_event_date'.
df_1['same_date'] = np.where(df_1['new_event_date']==df_1['new_event_date2'], True, False)
df_1.head(3)

index  event_date  year  month  day country_code  latitude  longitude  \
0      0  2022-11-04  2022     11    4           ES  41.51019    2.24589   
1      1  2021-08-27  2021      8   27           IT  44.40289    8.98775   
2      3  2022-08-11  2022      8   11           IT  41.70922   12.78512   

                                             witness  \
0                                       Roger Eritja   
1  Karin Bakran-Lebl;Ana Klobucar;UNIROMA1;Roger ...   
2  UNIROMA1;Eleonora Longo;Francesco Severini;Rog...   

                                issue  ...  prcp snow   wdir  wspd  wpgt  \
0  CONTINENT_DERIVED_FROM_COORDINATES  ...   0.0  NaN  271.0  19.5  42.6   
1  CONTINENT_DERIVED_FROM_COORDINATES  ...   0.3  NaN   68.0  13.8  27.8   
2  CONTINENT_DERIVED_FROM_COORDINATES  ...   1.1  NaN  173.0   9.0   NaN   

     pres  tsun  new_event_date  new_event_date2  same_date  
0  1013.8   NaN      2022-11-04       2022-11-04       True  
1  1007.5   NaN      2021-08-27       2021-08-27       True  
2  1013.9   NaN      2022-08-11       2022-08-11       True  

[3 rows x 28 columns]

In [99]:
# Check if all the rows have True as value in column 'same_date'.
df_1['same_date'].value_counts()

True    14512
Name: same_date, dtype: int64

All values are True, so all the rows have the same date on the same row.

In [100]:
# Check if column 'latitude' has the same information as 'latitude2'.
df_1['same_latitude'] = np.where(df_1['latitude']==df_1['latitude2'], True, False)
# Check if all the rows have True as value in column 'same_date'.
df_1['same_latitude'].value_counts()

True    14512
Name: same_latitude, dtype: int64

All values are True, so all the rows have the same latitude on the same row.

In [101]:
# Check if column 'latitude' has the same information as 'latitude2'.
df_1['same_longitude'] = np.where(df_1['longitude']==df_1['longitude2'], True, False)
# Check if all the rows have True as value in column 'same_date'.
df_1['same_longitude'].value_counts()

True    14512
Name: same_longitude, dtype: int64

All values are True, so all the rows have the same longitude on the same row.

### 1.2. Drop columns

In [102]:
df_2=df_1.copy()
df_2.columns

Index(['index', 'event_date', 'year', 'month', 'day', 'country_code',
       'latitude', 'longitude', 'witness', 'issue', 'index.1', 'event_date2',
       'latitude2', 'longitude2', 'time', 'tavg', 'tmin', 'tmax', 'prcp',
       'snow', 'wdir', 'wspd', 'wpgt', 'pres', 'tsun', 'new_event_date',
       'new_event_date2', 'same_date', 'same_latitude', 'same_longitude'],
      dtype='object')

In [103]:
df_1['issue'].value_counts()

CONTINENT_DERIVED_FROM_COORDINATES    14512
Name: issue, dtype: int64

In [104]:
df_1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 14512 entries, 0 to 14511
Data columns (total 30 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   index            14512 non-null  int64         
 1   event_date       14512 non-null  object        
 2   year             14512 non-null  int64         
 3   month            14512 non-null  int64         
 4   day              14512 non-null  int64         
 5   country_code     14512 non-null  object        
 6   latitude         14512 non-null  float64       
 7   longitude        14512 non-null  float64       
 8   witness          14512 non-null  object        
 9   issue            14512 non-null  object        
 10  index.1          14512 non-null  int64         
 11  event_date2      14512 non-null  object        
 12  latitude2        14512 non-null  float64       
 13  longitude2       14512 non-null  float64       
 14  time             12485 non-null  objec

We can see some interesting information about some columns:
- 'snow': only has 76 rows with data, so it can not be used to predict.
- 'wpgt': has 4882 rows with data, it will not be used to predict.
- 'tsun': only has 1 row with data, so it can not be used to predict.

In [105]:
# Create a list with the columns that must be deleted.
columns_to_delete=['index', 'event_date','event_date2','latitude2', 'longitude2','same_date','time','wpgt',
                     'same_date', 'same_latitude', 'same_longitude','new_event_date2','issue', 'snow','tsun']
df_2=df_2.drop(columns_to_delete,axis=1,inplace=False)
df_2.columns

Index(['year', 'month', 'day', 'country_code', 'latitude', 'longitude',
       'witness', 'index.1', 'tavg', 'tmin', 'tmax', 'prcp', 'wdir', 'wspd',
       'pres', 'new_event_date'],
      dtype='object')

### 1.3. Change columns names

In [106]:
df_3=df_2.copy()
df_3.columns

Index(['year', 'month', 'day', 'country_code', 'latitude', 'longitude',
       'witness', 'index.1', 'tavg', 'tmin', 'tmax', 'prcp', 'wdir', 'wspd',
       'pres', 'new_event_date'],
      dtype='object')

In [107]:
# Change columns names.
new_columns_names={'year':'year', 'month':'month','day':'day', 'country_code':'country_code',
                    'latitude':'latitude','longitude':'longitude', 'witness':'witness',
                     'tavg':'temp_avg', 'tmin':'temp_min','tmax':'temp_max', 'prcp':'precipitation',
                     'wdir':'wind_dir', 'wspd':'wind_speed','pres':'pressure','new_event_date':'event_date'}

In [108]:
df_3.rename(columns=new_columns_names,inplace=True)

# Check columns names now.
df_3.columns

Index(['year', 'month', 'day', 'country_code', 'latitude', 'longitude',
       'witness', 'index.1', 'temp_avg', 'temp_min', 'temp_max',
       'precipitation', 'wind_dir', 'wind_speed', 'pressure', 'event_date'],
      dtype='object')

In [109]:
df_3.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 14512 entries, 0 to 14511
Data columns (total 16 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   year           14512 non-null  int64         
 1   month          14512 non-null  int64         
 2   day            14512 non-null  int64         
 3   country_code   14512 non-null  object        
 4   latitude       14512 non-null  float64       
 5   longitude      14512 non-null  float64       
 6   witness        14512 non-null  object        
 7   index.1        14512 non-null  int64         
 8   temp_avg       12480 non-null  float64       
 9   temp_min       12454 non-null  float64       
 10  temp_max       12453 non-null  float64       
 11  precipitation  11072 non-null  float64       
 12  wind_dir       10404 non-null  float64       
 13  wind_speed     11992 non-null  float64       
 14  pressure       11384 non-null  float64       
 15  event_date     1451

### 1.4 Delete 'wind_dir' null values

Checking weather columns it can be seen that wind direction ('wind_dir') has 10404 non-null values, so it has more than 4000 null values. Those rows will be deleted because the dataframe still has enough rows.

In [110]:
null_wind_rows=df_3[df_3['wind_dir'].isnull()]
null_wind_rows.head(3)

year  month  day country_code  latitude  longitude  \
6   2022      6   22           IT  40.30343   17.72688   
14  2022      6   23           IT  44.83439   11.13635   
17  2021      8    5           IT  44.38320    7.53769   

                                              witness  index.1  temp_avg  \
6                               Roger Eritja;UNIROMA1        6       NaN   
14                              Roger Eritja;UNIROMA1       14       NaN   
17  Ana Klobucar;UNIROMA1;Hugo Costa Osório;Roger ...       17       NaN   

    temp_min  temp_max  precipitation  wind_dir  wind_speed  pressure  \
6        NaN       NaN            NaN       NaN         NaN       NaN   
14       NaN       NaN            NaN       NaN         NaN       NaN   
17       NaN       NaN            NaN       NaN         NaN       NaN   

   event_date  
6  2022-06-22  
14 2022-06-23  
17 2021-08-05

In [111]:
null_index=null_wind_rows.index
len(null_index)

4108

In [112]:
df_4=df_3.drop(null_index,axis=0,inplace=False)
df_4['wind_dir'].isnull().sum()

0

In [113]:
print(f'This DataFrame has {df_4.shape[0]} rows and {df_4.shape[1]} columns.')

This DataFrame has 10404 rows and 16 columns.


## 2. Get location information

To get the number of mosquito per postal code geopy library will be used.

In [151]:
from geopy.geocoders import Nominatim
from geopy.exc import GeocoderTimedOut
geolocator = Nominatim(user_agent="new")
from tqdm import tqdm
tqdm.pandas()
import time
from time import sleep
import requests

In [115]:
len(df_4['country_code'].unique())

38

In [116]:
df_test1=df_4.head(10)
df_test1

year  month  day country_code  latitude  longitude  \
0   2022     11    4           ES  41.51019    2.24589   
1   2021      8   27           IT  44.40289    8.98775   
2   2022      8   11           IT  41.70922   12.78512   
3   2022      6   27           HU  47.46323   19.17698   
4   2021     10   20           ES  39.43645    2.75412   
5   2022      8   26           IT  43.76617   11.27971   
7   2022      8   16           ES  37.39345   -6.08236   
8   2022      8   13           IT  45.57203   12.10044   
9   2022      6   22           IT  41.83784   12.69665   
10  2022      6   26           HU  47.47027   19.04295   

                                              witness  index.1  temp_avg  \
0                                        Roger Eritja        0      15.4   
1   Karin Bakran-Lebl;Ana Klobucar;UNIROMA1;Roger ...        1      25.0   
2   UNIROMA1;Eleonora Longo;Francesco Severini;Rog...        2      27.7   
3   Anonymous expert;Kornélia Kurucz;Gábor Kemenes...        3      26.6   
4   Maria Angeles Puig;Mikel Bengoa Paulis;Ignacio...        4      18.3   
5                               UNIROMA1;Roger Eritja        5      26.1   
7   Isis Sanpera-Calbet;Mikel Alexander González;D...        7      24.9   
8   UNIROMA1;Eleonora Longo;Francesco Severini;Rog...        8      23.1   
9                               Roger Eritja;UNIROMA1        9      27.5   
10  Kornélia Kurucz;Anonymous expert;Gábor Kemenes...       10      25.2   

    temp_min  temp_max  precipitation  wind_dir  wind_speed  pressure  \
0       11.0      20.0            0.0     271.0        19.5    1013.8   
1       23.0      28.1            0.3      68.0        13.8    1007.5   
2       22.0      32.0            1.1     173.0         9.0    1013.9   
3       17.7      33.1            0.0      89.0         6.2    1016.6   
4       17.7      24.8            0.0      83.0         7.1    1023.5   
5       19.9      33.3            0.3      34.0         5.4    1010.9   
7       18.7      30.9            0.0     256.0        14.2    1011.3   
8       16.0      28.3            0.0      31.0         7.2    1013.0   
9       20.1      33.8            0.0     176.0        10.1    1014.0   
10      18.7      31.5            0.0     277.0         7.2    1014.8   

   event_date  
0  2022-11-04  
1  2021-08-27  
2  2022-08-11  
3  2022-06-27  
4  2021-10-20  
5  2022-08-26  
7  2022-08-16  
8  2022-08-13  
9  2022-06-22  
10 2022-06-26

In [117]:
def get_country_short(lat,lon):
    pos = str(lat) + ', ' + str(lon)
    locations = geolocator.reverse(pos)
    return locations.raw#['address']['postcode']

In [118]:
get_country_short(47.46323,19.17698)


{'place_id': 87535041,
 'licence': 'Data © OpenStreetMap contributors, ODbL 1.0. http://osm.org/copyright',
 'osm_type': 'way',
 'osm_id': 311983099,
 'lat': '47.4631765',
 'lon': '19.177015822170688',
 'class': 'building',
 'type': 'yes',
 'place_rank': 30,
 'importance': 9.99999999995449e-06,
 'addresstype': 'building',
 'name': '',
 'display_name': '55, Zsolt utca, Erzsébettelep, XVIII. kerület, Budapest, Közép-Magyarország, 1183, Magyarország',
 'address': {'house_number': '55',
  'road': 'Zsolt utca',
  'suburb': 'Erzsébettelep',
  'district': 'XVIII. kerület',
  'city': 'Budapest',
  'ISO3166-2-lvl6': 'HU-BU',
  'region': 'Közép-Magyarország',
  'postcode': '1183',
  'country': 'Magyarország',
  'country_code': 'hu'},
 'boundingbox': ['47.4631165', '47.4632365', '19.1769142', '19.1771174']}

In [119]:
get_country_short(40.30343,17.72688)

{'place_id': 80846324,
 'licence': 'Data © OpenStreetMap contributors, ODbL 1.0. http://osm.org/copyright',
 'osm_type': 'way',
 'osm_id': 498502819,
 'lat': '40.303382996015685',
 'lon': '17.726909989184243',
 'class': 'highway',
 'type': 'secondary',
 'place_rank': 26,
 'importance': 0.10000999999999993,
 'addresstype': 'road',
 'name': 'Via del Mare',
 'display_name': 'Via del Mare, Quota 10, Torre Colimena, Manduria, Taranto, Puglia, Italia',
 'address': {'road': 'Via del Mare',
  'neighbourhood': 'Quota 10',
  'village': 'Torre Colimena',
  'town': 'Manduria',
  'county': 'Taranto',
  'ISO3166-2-lvl6': 'IT-TA',
  'state': 'Puglia',
  'ISO3166-2-lvl4': 'IT-75',
  'country': 'Italia',
  'country_code': 'it'},
 'boundingbox': ['40.3019054', '40.3047878', '17.7245981', '17.7439692']}

In [164]:
def get_country(row):
    pos = str(row['latitude']) + ', ' + str(row['longitude'])
    locations = geolocator.reverse(pos)
    try:
        sleep(0.01)
        return locations.raw['address']['postcode']
    except:
        sleep(0.01)
        return np.nan

In [165]:
df_test1['post_code'] = df_test1.apply(lambda row: get_country(row), axis = 1)

/tmp/ipykernel_9229/3876475917.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_test1['post_code'] = df_test1.apply(lambda row: get_country(row), axis = 1)


In [122]:
df_test1

year  month  day country_code  latitude  longitude  \
0   2022     11    4           ES  41.51019    2.24589   
1   2021      8   27           IT  44.40289    8.98775   
2   2022      8   11           IT  41.70922   12.78512   
3   2022      6   27           HU  47.46323   19.17698   
4   2021     10   20           ES  39.43645    2.75412   
5   2022      8   26           IT  43.76617   11.27971   
7   2022      8   16           ES  37.39345   -6.08236   
8   2022      8   13           IT  45.57203   12.10044   
9   2022      6   22           IT  41.83784   12.69665   
10  2022      6   26           HU  47.47027   19.04295   

                                              witness  index.1  temp_avg  \
0                                        Roger Eritja        0      15.4   
1   Karin Bakran-Lebl;Ana Klobucar;UNIROMA1;Roger ...        1      25.0   
2   UNIROMA1;Eleonora Longo;Francesco Severini;Rog...        2      27.7   
3   Anonymous expert;Kornélia Kurucz;Gábor Kemenes...        3      26.6   
4   Maria Angeles Puig;Mikel Bengoa Paulis;Ignacio...        4      18.3   
5                               UNIROMA1;Roger Eritja        5      26.1   
7   Isis Sanpera-Calbet;Mikel Alexander González;D...        7      24.9   
8   UNIROMA1;Eleonora Longo;Francesco Severini;Rog...        8      23.1   
9                               Roger Eritja;UNIROMA1        9      27.5   
10  Kornélia Kurucz;Anonymous expert;Gábor Kemenes...       10      25.2   

    temp_min  temp_max  precipitation  wind_dir  wind_speed  pressure  \
0       11.0      20.0            0.0     271.0        19.5    1013.8   
1       23.0      28.1            0.3      68.0        13.8    1007.5   
2       22.0      32.0            1.1     173.0         9.0    1013.9   
3       17.7      33.1            0.0      89.0         6.2    1016.6   
4       17.7      24.8            0.0      83.0         7.1    1023.5   
5       19.9      33.3            0.3      34.0         5.4    1010.9   
7       18.7      30.9            0.0     256.0        14.2    1011.3   
8       16.0      28.3            0.0      31.0         7.2    1013.0   
9       20.1      33.8            0.0     176.0        10.1    1014.0   
10      18.7      31.5            0.0     277.0         7.2    1014.8   

   event_date post_code  
0  2022-11-04     08105  
1  2021-08-27     16147  
2  2022-08-11     00049  
3  2022-06-27      1183  
4  2021-10-20     07609  
5  2022-08-26     50136  
7  2022-08-16     41960  
8  2022-08-13     30033  
9  2022-06-22     00044  
10 2022-06-26      1119

In [124]:
df_4.shape

(10404, 16)

In [126]:
df4_1 = df_4.iloc[0:1000]
df4_2 = df_4.iloc[1000:2000]
df4_3 = df_4.iloc[2000:3000]
df4_4 = df_4.iloc[3000:4000]
df4_5 = df_4.iloc[4000:5000]
df4_6 = df_4.iloc[5000:6000]
df4_7 = df_4.iloc[6000:7000]
df4_8 = df_4.iloc[7000:8000]
df4_9 = df_4.iloc[8000:9000]
df4_10 = df_4.iloc[9000:10500]

In [167]:
df4_10['country'] = df4_10.apply(lambda row: get_country(row), axis = 1)
time.sleep(1)

/tmp/ipykernel_9229/2259268977.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df4_10['country'] = df4_10.apply(lambda row: get_country(row), axis = 1)


## 3. Analyze weather data

In [ ]:
df_3.columns

Our present dataframe has the following columns:
- **year**: year of mosquito report.

- **month**: month of mosquito report.

- **day**: day of mosquito report.

- **country_code**: country where the mosquito was located.

- **latitude**: latitude of mosquito report.

- **longitude**: longitude of mosquito report.

- **witness**: people that saw the mosquito.

- **temp_avg**: average air temperature in °C.

- **temp_min**: minimum air temperature in °C.

- **temp_max**: maximum air temperature in °C.

- **precipitation**: precipitation total in mm.

- **wind_dir**: average wind direction in degrees (°).

- **wind_speed**: average wind speed in km/h.

- **pressure**: average sea-level air pressure in hPa.

- **event_date**: date of mosquito report.

In [ ]:
# Main statistics from numeric columns.
df_3.describe().T

This table shows:
- **year**:
    - the first year of report is 2014.
    - the last year of report is 2022.
    - the mean year is 2019 but 50% of data are at 2020.
- **month**:
    - the first month of report is 1 (January).
    - the last month of report is 12 (December).
    - the mean month and 50% of data are at month 8 (August).
- **day**:
    - the first day of report is 1st.
    - the last day of report is 31st.
    - the mean day is 16th.
- **latitude** and **latitude**: there is no use in analizing them, they are coordinates.
- **temp_avg**:
    - the lowest average air temperature is -0.9°C.
    - the highest average air temperature is 34.8°C.
    - the mean average air temperature and 50% of data are close to 23°C.
- **temp_min**:
    - the lowest minimum air temperature is -3.5°C.
    - the highest minimum air temperature is 29.0°C.
    - the mean minimum air temperature and 50% of data are close to 18°C.
- **temp_max**:
    - the lowest maximum air temperature is 1.2°C.
    - the highest maximum air temperature is 50.0°C.
    - the mean maximum air temperature and 50% of data are close to 28°C.
- **precipitation**:
    - the lowest precipitation is 0mm.
    - the highest precipitation is 179.3mm.
    - the mean precipitation and 50% of data are close to 0mm.
- **wind_dir**:
    - the lowest wind direction degrees is 0° (North).
    - the highest wind direction degrees is 360° (North).
    - the mean wind direction degrees and 50% of data are close to 170°(Southeast).
- **wind_dir**:
    - the lowest wind direction degrees is 0° (North).
    - the highest wind direction degrees is 360° (North).
    - the mean wind direction degrees and 50% of data are close to 170°(Southeast).
- **pressure**:
    - the lowest pressure is 993hPa.
    - the highest pressure is 1033hPa.
    - the mean pressure and 50% of data are close to 1016hPa.

In [ ]:
# Main statistics from categorical columns.
df_3.describe(include='object').T

This table shows:
- **country_code**:
    - country_code has 46 unique values.
    - the most frequent country is ES (Spain).
- **witness**:
    - witness has 2034 unique values.
    - the most frequent witnesses are 'UNIROMA1;Roger Eritja'.

In [ ]:
# Check null values.
df_3.isnull().sum()

In [ ]:
# Check for outliers on the weather data.
weather_columns=['temp_avg', 'temp_min', 'temp_max', 'precipitation','wind_dir', 'wind_speed', 'pressure']

In [ ]:
weather_cols=['temp_avg', 'temp_min', 'temp_max', 'precipitation','wind_dir', 'wind_speed', 'pressure']
df_w=df_3[weather_cols]
df_w

### 3.1. Outliers

In [ ]:
# Set figure characteristics.
fig, axes =plt.subplots(4,2,figsize=(15,10))

# Flatter axes to plot them easier/faster using a for loop.
axes = axes.flat

for i, col in enumerate(df_w.columns):
    sns.boxplot(x = df_w[col], data = df_w, ax=axes[i], color = "violet")

# Delete empty graph.
fig.delaxes(axes[7])

# Adjust graphs so they don't overlap.
fig.tight_layout()

# Show figures.
plt.show();

This graphs show that there are outliers in every columns excepto for wind_dir. they also show where outliers are located in each variable.

In [ ]:
def get_outliers(columns_list, dataframe):
    dicc_indices = {} # create a dictionary to store indexes.

    # iterate on the weather columns list
    for col in columns_list:
        #get quartiles Q1 and Q3
        Q1 = np.nanpercentile(dataframe[col], 25)
        Q3 = np.nanpercentile(dataframe[col], 75)

        # get interquartile range (IQR)
        IQR = Q3 - Q1

        # get limits
        outlier_step = 1.5 * IQR

        # filtrate our dataframe to identify outliers
        outliers_data = dataframe[(dataframe[col] < Q1 - outlier_step) | (dataframe[col] > Q3 + outlier_step)]

        if outliers_data.shape[0] > 0: # check if our dataframe has outliers
            dicc_indices[col] = (list(outliers_data.index))
    return dicc_indices

In [ ]:
# get columns with outliers.
ind_out=get_outliers(weather_cols, df_3)
ind_out.keys()

### 3.2. Duplicates

Duplicates were already checked at the beggining of the project. Now it makes no sense to check for them because new rows have not been added and false duplicates can be found due to lack of some columns that added some distinguishing informtation such as 'gbifID.

In [ ]:
# Check this posible mistake checking for duplicates now.
df_3.duplicated().sum()

240 duplicated rows were found, but if gbifID had been kept it would show that all the row has the same data except for gbifID because they were different mosquitos observed at the same time.

In [ ]:
# Get a dataframe with only duplicated rows.
duplicates_data = df_3[(df_3.duplicated())]
duplicates_data.shape

In [ ]:
# Look for for the most repeated date.
duplicates_data['event_date'].value_counts()

In [ ]:
# Check rows with that date to see the rest of the data.
duplicates_data[duplicates_data['event_date']=='2021-09-11']

In [ ]:
# Check the first dataframe used in this project in code1 to see the other columns that were deleted.
df_check_dupli = pd.read_csv('../data/gbif_mosquito_data.csv', sep='\t')
df_check_dupli.columns

In [ ]:
# Look at the same rows with the same date as before to see the rest of the information.
df_check_dupli[df_check_dupli['eventDate']=='2021-09-11T00:00:00'].sort_values('decimalLatitude')

Looking at the first dataframe that was used in this project it can be seen that those rows that seemed to be duplicated now have different 'gbifID' and different 'occurrenceID', so they are different mosquitoes observed.

In [ ]:
# Set figure characteristics.
fig, axes = plt.subplots(1, 7, figsize=(20,7))

# Create a for loop to creat one violinplot for each numerical variable.
for i in range(len(weather_columns)): 
    sns.violinplot(x='year', y=weather_columns[i], data = df_3 , ax = axes[i], palette = "rocket")   

# Adjust graphs so they don't overlap.
fig.tight_layout()

# Show figure.
plt.show()

In [ ]:
# Check for outliers in month using another method.
#mean_month=df_1['month'].mean()
#std_month=df_1['month'].std()
#upper=mean_month + std_month
#lower=mean_month - std_month
#ucb=mean_month + std_month * 3
#lcb=mean_month - std_month * 3

In [ ]:
## Start the graph.
#month_graph=sns.histplot(x=df_1['month'],kde=True)
#graph.axvline(x=mean_month, c='red',label='mean')
#
## Plot its standard deviation.
#graph.axvline(x=upper,c='green',label='std')
#graph.axvline(x=lower, c='green')
#
## Plot its confidence interval to 99.7%
#graph.axvline(x=lcb, c='orange',label='99 lower')
#graph.axvline(x=ucb, c='orange',label='99 upper')
#
#plt.legend()